cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -49,89 +131,97 @@ void CUDA_DEVICE_CALLABLE uint32ToLowercaseHexString(uint32_t num, char* destina
-    uint32_t D = hash_state->hash_value[3];
-
-    for (unsigned int j = 0; j < 64; j++) {
-      uint32_t F;
-      uint32_t g;
-      switch (j / 16) {
-        case 0:
-          F = (B & C) | ((~B) & D);
-          g = j;
-          break;
-        case 1:
-          F = (D & B) | ((~D) & C);
-          g = (5 * j + 1) % 16;
-          break;
-        case 2:
-          F = B ^ C ^ D;
-          g = (3 * j + 5) % 16;
-          break;
-        case 3:
-          F = C ^ (B | (~D));
-          g = (7 * j) % 16;
-          break;
-      }
-
-      uint32_t buffer_element_as_int;
-      std::memcpy(&buffer_element_as_int, hash_state->buffer + g * 4, 4);
-      F = F + A + md5_hash_constants[j] + buffer_element_as_int;
-      A = D;
-      D = C;
-      C = B;
-      B = B + __funnelshift_l(F, F, md5_shift_constants[((j / 16) * 4) + (j % 4)]);
-    }
-
-    hash_state->hash_value[0] += A;
-    hash_state->hash_value[1] += B;
-    hash_state->hash_value[2] += C;
-    hash_state->hash_value[3] += D;
-
-    hash_state->buffer_length = 0;
+    release_assert(false && "MD5 Unsupported chrono type column");
   }
 
-  /**
-   * @brief Core MD5 element processing function
-   */
-  template <typename TKey>
-  void __device__ process(TKey const& key, md5_intermediate_data* hash_state) const
+  template <typename T, typename std::enable_if_t<!is_fixed_width<T>()>* = nullptr>
+  void __device__ operator()(column_device_view data_col,
+                             size_type offset_begin,
+                             size_type offset_end,
+                             md5_intermediate_data* hash_state) const
   {
-    uint32_t const len  = sizeof(TKey);
-    uint8_t const* data = reinterpret_cast<uint8_t const*>(&key);
-    hash_state->message_length += len;
+    release_assert(false && "MD5 Unsupported non-fixed-width type column");
+  }
 
-    // 64 bytes for the number of bytes processed in a given step
-    constexpr int md5_chunk_size = 64;
-    if (hash_state->buffer_length + len < md5_chunk_size) {
-      thrust::copy_n(thrust::seq, data, len, hash_state->buffer + hash_state->buffer_length);
-      hash_state->buffer_length += len;
-    } else {
-      uint32_t copylen = md5_chunk_size - hash_state->buffer_length;
+  template <typename T, typename std::enable_if_t<is_floating_point<T>()>* = nullptr>
+  void __device__ operator()(column_device_view data_col,
+                             size_type offset_begin,
+                             size_type offset_end,
+                             md5_intermediate_data* hash_state) const
+  {
+    for (int i = offset_begin; i < offset_end; i++) {
+      if (!data_col.is_null(i)) {
+        T const& key = data_col.element<T>(i);
+        if (isnan(key)) {
+          T nan = std::numeric_limits<T>::quiet_NaN();
+          md5_process(nan, hash_state);
+        } else if (key == T{0.0}) {
+          md5_process(T{0.0}, hash_state);
+        } else {
+          md5_process(key, hash_state);
+        }
```

Comment:
Would be nice to have a helper like `normalize_nans_and_zeros_lambda` that we can call both here and in `MD5Hash::operator()`, instead of repeating the logic.

Reply:
Would a straight helper function be appropriate so I can remove the repeating logic from murmur hash too?

Reply:
Which one would you prefer?
```
template <typename T>
struct md5_normalize_nans_and_zeros_lambda {
  column_device_view data_col;
  md5_intermediate_data* hash_state;
  __device__ void operator()(int index)
  {
    T const& key = data_col.element<T>(index);
    if (isnan(key)) {
      T nan = std::numeric_limits<T>::quiet_NaN();
      md5_process(nan, hash_state);
    } else if (key == T{0.0}) {
      md5_process(T{0.0}, hash_state);
    } else {
      md5_process(key, hash_state);
    }
  }
};
```
vs.
```
template <typename TKey>
TKey CUDA_DEVICE_CALLABLE normalize_nans_and_zeros(TKey key) {
  if (isnan(key)) {
    return std::numeric_limits<TKey>::quiet_NaN();
  } else if (key == TKey{0.0}) {
    return TKey{0.0};
  } else {
    return key;
  }
}
```

Reply:
I was thinking of the second option, but haven't looked into it too much. Whichever you think works best :)

Reply:
Went with the second option, it is more reusable in future hash functions than the first. However, did not use it in the murmur hash normalization step, which uses a slightly different logic when it finds a `0.0` value.